packages/autocomplete-core/src/__tests__/concurrency.test.ts

```diff
@@ -0,0 +1,58 @@
+import userEvent from '@testing-library/user-event';
+
+import { createAutocomplete } from '../createAutocomplete';
+
+function defer<TValue>(fn: () => TValue, timeout: number) {
+  return new Promise<TValue>((resolve) => {
+    setTimeout(() => resolve(fn()), timeout);
+  });
+}
+
+describe('concurrency', () => {
+  test('resolves the responses in order from getSources', async () => {
+    // These delays make the second query come back after the third one.
+    const delays = [100, 300, 200];
+    let deferCount = -1;
+
+    const getSources = ({ query }) => {
+      deferCount++;
+
+      return defer(() => {
+        return [
+          {
+            getItems() {
+              return [{ label: query }];
+            },
+          },
+        ];
+      }, delays[deferCount]);
+    };
+    const onStateChange = jest.fn();
+    const autocomplete = createAutocomplete({ getSources, onStateChange });
+    const { onChange } = autocomplete.getInputProps({ inputElement: null });
+    const input = document.createElement('input');
+    input.addEventListener('input', onChange);
+    document.body.appendChild(input);
+
+    userEvent.type(input, 'a');
+    userEvent.type(input, 'b');
+    userEvent.type(input, 'c');
+
+    await defer(() => {}, Math.max(...delays));
+
+    const itemsHistory: Array<{ label: string }> = (onStateChange.mock
+      .calls as any).flatMap((x) =>
+      x[0].state.collections.flatMap((x) => x.items)
+    );
+
+    // The second query should never have brought results.
+    expect(itemsHistory.find((x) => x.label === 'ab')).toBeUndefined();
+
+    // The last item must be the one corresponding to the last query.
+    expect(itemsHistory[itemsHistory.length - 1]).toEqual(
+      expect.objectContaining({ label: 'abc' })
+    );
```

Comment:
Maybe add a test to check for the presence of `{ label: 'a' }` as well as the first item?

Reply:
maybe a snapshot of all the labels could make sense?

Reply:
A snapshot will be fragile because there's no certainty of how many times `onStateChange` is called. It's called whenever a setter is called (e.g. `setIsOpen`, etc.), and a setter can be called user-land. It means that if we refactor something, this test will become a false negative.

I'll check for the presence of `a` though.